frontend/plan/reducers/schedule.js

```diff
@@ -154,246 +76,148 @@ const nextAvailable = (scheduleName, used) => {
         ...state,
-        deletedSchedules: localOnly
-            ? state.deletedSchedules
-            : {
-                  ...(state.deletedSchedules || {}),
-                  [state.schedules[scheduleName].id]: {
-                      deletionQueued: false,
-                      attempts: 0,
-                  },
-              },
         schedules: newSchedules,
         scheduleSelected:
             scheduleName === state.scheduleSelected
                 ? Object.keys(newSchedules)[0]
                 : state.scheduleSelected,
+        deletedSchedules: state.deletedSchedules.filter(
+            (deletedScheduleName) => deletedScheduleName !== scheduleName
+        ),
     };
 };
 
 /**
- * Returns whether the given array of meetings/sections contains
- * courses not from the current semester.
- * @param meetings
- * @param currentSemester
+ * Takes in schedules from the backend; incorporates them into state; and returns the new state
+ * @param state The redux state
+ * @param schedulesFromBackend The schedules object sent by the backend
+ * @returns {{cartSections: *, cartPushedToBackend: boolean, schedules: *}}
  */
-const containsOldSemester = (meetings, currentSemester) =>
-    meetings.reduce(
-        (acc, { semester }) =>
-            acc || (semester && semester !== currentSemester),
-        false
-    );
+const handleUpdateSchedulesOnFrontend = (state, schedulesFromBackend) => {
+    let newState = { ...state };
+
+    schedulesFromBackend.forEach((scheduleFromBackend) => {
+        if (
+            newState.deletedSchedules.reduce(
+                (acc, schedule) => acc || schedule === scheduleFromBackend.name,
+                false
+            )
+        ) {
+            return;
+        }
+
+        const cloudUpdated = new Date(scheduleFromBackend.updated_at).getTime();
+
+        if (scheduleFromBackend.name === "cart") {
+            // If changes to the cart are still syncing, ignore the requested update
+            if (
+                newState.cartPushedToBackend &&
+                cloudUpdated >= newState.cartUpdatedAt &&
+                cloudUpdated - newState.cartUpdatedAt >= MIN_TIME_DIFFERENCE
+            ) {
+                newState = {
+                    ...newState,
+                    cartId: scheduleFromBackend.id,
+                    cartSections: scheduleFromBackend.sections,
+                    cartUpdatedAt: scheduleFromBackend.updated_at,
+                };
+            }
+        } else {
+            const foundSchedule = newState.schedules[scheduleFromBackend.name];
+            // If changes to the schedule are still syncing, ignore the requested update
+            if (
+                !foundSchedule ||
+                (foundSchedule &&
+                    foundSchedule.pushedToBackend &&
+                    cloudUpdated >= foundSchedule.updated_at &&
+                    foundSchedule.updated_at - cloudUpdated >=
+                        MIN_TIME_DIFFERENCE)
```

Comment:
`foundSchedule.updated_at - cloudUpdated` would always be negative and thus this if statement would always be `false`?